drivers/modem/gsm_ppp.c

```diff
@@ -612,6 +657,11 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		goto attaching;
 	}
 
+	/* If modem is searching for network, we should skip the setup step */
+	if (gsm->net_state == GSM_SEARCHING) {
```

Comment:
@endian-benjamin following your comments I do spot an issue here and I think this should be extended to include `GSM_NOT_REGISTERED`, `GSM_REGISTRATION_DENIED` & `GSM_UNKNOWN` as well

Reply:
Wat. That makes no sense.

Why not just reset state to NOT searching if `AT+CREG?` fails?

Reply:
> Why not just reset state to NOT searching if `AT+CREG?` fails?

I don't think that we should reset it to `NOT_REGISTERED` if it is indeed searching for network, it just means that we should wait longer in this state.

I'm proposing the following:
```c
	if (gsm->net_state != GSM_ROAMING && gsm->net_state != GSM_HOME_NETWORK) {
		goto registering;
	}
```

The registration step below wants `GSM_ROAMING` / `GSM_HOME_NETWORK` before moving on to the next step. So we just make sure that we query the registration again as long as the modem is not registered at a cell tower.

Reply:
That's what other modems are doing:
- [simcom](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/simcom-sim7080.c#L1154)
- [ublox sara](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/ublox-sara-r4.c#L1297)

Reply:
I'm sure you have basically the right logic in mind, but the code your proposing seems very unintuitive.